onnxruntime/core/providers/webnn/builders/impl/gemm_op_builder.cc

```diff
@@ -38,7 +38,16 @@ Status GemmOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const N
   emscripten::val b = model_builder.GetOperand(node.InputDefs()[b_idx]->Name());
   emscripten::val output = emscripten::val::object();
   if (op_type == "MatMul") {
-    output = model_builder.GetBuilder().call<emscripten::val>("matmul", a, b);
+    std::vector<int64_t> a_shape;
+    if (!GetShape(*input_defs[a_idx], a_shape, logger)) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Can not get shape of A.");
+    }
+    // The inputs of matmul must be at least 3D for WebNN CPU EP, use gemm for 2D case.
```

Comment:
```suggestion
    // The inputs of MatMul must be at least 3D for WebNN CPU EP. Use GEMM for 2D case.
```

Though, the proper fix for this belongs in the Chromium XNNPack backend, not as a workaround in ORT. From the spec, MatMul expects 2D inputs (https://www.w3.org/TR/webnn/#api-mlgraphbuilder-matmul), and so the Chromium XNNPack backend should just reshape the 2D input (no copy needed) by inserting a leading 1. That way every client (e.g. TensorFlow.js too) doesn't need to repeat the same workaround.

Reply:
Thanks @fdwr, totally agree. I will tell @lisa0314 to drop a CL to Chromium.

@zesongw, let's remove this workaround in WebNN EP.

Reply:
XNNPACK reshape is implemented by a copy operator: [third_party/xnnpack/src/src/subgraph/static-reshape.c](https://source.chromium.org/chromium/chromium/src/+/main:third_party/xnnpack/src/src/subgraph/static-reshape.c;l=249;drc=f4a00cc248dd2dc8ec8759fb51620d47b5114090;bpv=1;bpt=1). Inserting reshape would introduce extra perf overhead.

Reply:
> XNNPACK reshape is implemented by a copy operator: [third_party/xnnpack/src/src/subgraph/static-reshape.c](https://source.chromium.org/chromium/chromium/src/+/main:third_party/xnnpack/src/src/subgraph/static-reshape.c;l=249;drc=f4a00cc248dd2dc8ec8759fb51620d47b5114090;bpv=1;bpt=1). Inserting reshape would introduce extra perf overhead.

Oof, XNNPack has no copyless reshape? 😢 It should just be a matter of using different dimensions/strides over the data without copying the entire tensor data, and it would be best for all WebNN implementations to treat `reshape` as "basically free" so they can insert them freely as needed for compat between them and WebNN, but that does strengthen the need for the implementation to be in Chromium rather than ORT.